### PR TITLE
IGNITE-17148 Support for abstract configuration

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-17148

### CODE REVIEW

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java [hunk 1]
```diff
@@ -3406,4 +3425,21 @@ private static void addInternalSchemaTypesMethod(
                 .append(getThisFieldCode(mtd, internalSchemaTypesFieldDef))
                 .retObject();
     }
+
+    /**
+     * Adds an override for the {@link InnerNode#isExtendAbstractConfiguration()} method that returns {@code true}.
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java:3430]: `isExtendAbstractConfiguration` is not a good name from the grammatical perspective. It should either be ` extendsAbstractConfiguration` (which I prefer) or `doesExtendAbstractConfiguration`

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java [hunk 2]
```diff
@@ -3406,4 +3425,21 @@ private static void addInternalSchemaTypesMethod(
...
+        MethodDefinition mtd = innerNodeClassDef.declareMethod(
+                of(PUBLIC),
+                "isExtendAbstractConfiguration",
+                type(boolean.class)
+        );
+
+        mtd.getBody()
+                .push(true)
+                .retBoolean();
+    }
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java:3444]: This class is now 3444 lines long, holy Jesus

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/tree/InnerNode.java [hunk 3]
```diff
@@ -245,4 +246,11 @@ public Class<?>[] internalSchemaTypes() {
     public boolean isPolymorphic() {
         return false;
     }
+
+    /**
+     * Returns {@code true} if the config schema extends {@link AbstractConfiguration}.
+     */
+    public boolean isExtendAbstractConfiguration() {
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/tree/InnerNode.java:253]: please see my comment about this method's name

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java [hunk 4]
```diff
@@ -1149,4 +1149,13 @@ public static <T extends ConfigurationProperty<?>> T getByInternalId(NamedConfig
     public static <N> N getByInternalId(NamedListView<N> node, UUID internalId) {
         return node.get(((NamedListNode<?>) node).keyByInternalId(internalId));
     }
+
+    /**
+     * Checks whether configuration schema contains {@link AbstractConfiguration}.
+     *
+     * @param schemaClass Configuration schema class.
+     */
+    public static boolean isAbstractConfiguration(Class<?> schemaClass) {
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:1158]: Do we really need this method? It doesn't make the code shorter nor more readable

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java [hunk 5]
```diff
@@ -630,7 +631,6 @@ private static Map<Class<?>, Set<Class<?>>> schemaExtensions(
      * PolymorphicId}.
      *
      * @param schemaClass Configuration schema class.
-     * @return Schema fields.
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:633]: why did you remove this tag?

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 6]
```diff
@@ -204,6 +204,68 @@ public boolean contains(Object o) {
...
 
+    /**
+     * Union lists.
+     *
+     * @param lists Lists.
+     * @param <T> Type of the elements of lists.
+     * @return Immutable union of lists.
+     */
+    @SafeVarargs
+    public static <T> List<T> union(List<T>... lists) {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:215]: This is technically not a union. I think a better name would be `concat`

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 7]
```diff
@@ -204,6 +204,68 @@ public boolean contains(Object o) {
...
+        if (lists == null || lists.length == 0) {
+            return List.of();
+        }
+
+        return new AbstractList<>() {
+            /** {@inheritDoc} */
+            @Override
+            public T get(int index) {
+                for (int i = 0; i < lists.length; i++) {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:224]: `for (List<T> list : lists)`?

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 8]
```diff
@@ -204,6 +204,68 @@ public boolean contains(Object o) {
...
+            public Iterator<T> iterator() {
+                return concat(lists).iterator();
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public int size() {
+                int size = 0;
+
+                for (int i = 0; i < lists.length; i++) {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:248]: `for (List<T> list : lists)`?

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 9]
```diff
@@ -204,6 +204,68 @@ public boolean contains(Object o) {
...
+                    size += lists[i].size();
+                }
+
+                return size;
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public boolean contains(Object o) {
+                for (int i = 0; i < lists.length; i++) {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:258]: `for (List<T> list : lists)`?

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 10]
```diff
@@ -204,6 +204,68 @@ public boolean contains(Object o) {
...
+                    size += lists[i].size();
+                }
+
+                return size;
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public boolean contains(Object o) {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:257]: Why do you need to implement this method?

File: modules/core/src/test/java/org/apache/ignite/internal/util/CollectionUtilsTest.java [hunk 11]
```diff
@@ -269,6 +269,58 @@ void testConcatCollectionOfIterators() {
         assertEquals(List.of(1, 2, 3), collect(concat(List.of(List.of(1).iterator(), List.of(2, 3).iterator()))));
     }
 
+    @Test
+    void testListUnion() {
+        assertTrue(union(new List[0]).isEmpty());
+        assertTrue(union((List<Object>[]) null).isEmpty());
+        assertTrue(union(List.of()).isEmpty());
+
+        assertEquals(List.of(1), collect(union(List.of(1), List.of())));
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/internal/util/CollectionUtilsTest.java:278]: What are these `collect` calls for?

File: modules/configuration-annotation-processor/src/integrationTest/java/org/apache/ignite/internal/configuration/processor/ItConfigurationProcessorTest.java [hunk 12]
```diff
@@ -427,11 +432,186 @@ void testSuccessInternalIdFieldCodeGeneration() {
         assertTrue(batchCompile.getBySchema(cls).allGenerated());
     }
 
+    @Test
+    void testFailValidationForAbstractConfiguration() {
+        String packageName = "org.apache.ignite.internal.configuration.processor.abstractconfig.validation";
+
+        // Let's check the incompatible configuration schema annotations.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/integrationTest/java/org/apache/ignite/internal/configuration/processor/ItConfigurationProcessorTest.java:439]: I think that it is not necessary, or you insist?

File: modules/configuration-annotation-processor/src/integrationTest/resources/org/apache/ignite/internal/configuration/processor/abstractconfig/validation/MustNotBeSuperClassConfigurationSchema.java [hunk 13]
```diff
@@ -15,17 +15,14 @@
...
 
-package org.apache.ignite.internal.configuration.processor;
+package org.apache.ignite.internal.configuration.processor.abstractconfig.validation;
+
+import java.util.ArrayList;
+import org.apache.ignite.configuration.annotation.AbstractConfiguration;
 
 /**
- * Annotation processing exception.
+ * Checks that there should not be a superclass.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/integrationTest/resources/org/apache/ignite/internal/configuration/processor/abstractconfig/validation/MustNotBeSuperClassConfigurationSchema.java:24]: Yes, I tried changing the description.

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java [hunk 14]
```diff
@@ -365,23 +373,55 @@ private void createPojoBindings(
...
-        if (extendBaseSchema) {
-            DeclaredType superClassType = (DeclaredType) realSchemaClass.getSuperclass();
+        TypeElement superClass = superClass(realSchemaClass);
+
+        boolean isSuperClassAbstractConfiguration = !isClass(superClass.asType(), Object.class)
+                && superClass.getAnnotation(AbstractConfiguration.class) != null;
+
+        if (extendBaseSchema || isSuperClassAbstractConfiguration) {
+            DeclaredType superClassType = (DeclaredType) superClass.asType();
             ClassName superClassSchemaClassName = ClassName.get((TypeElement) superClassType.asElement());
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java:390]: Fix it

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java [hunk 15]
```diff
@@ -1051,21 +1060,151 @@ private void validateNameFields(TypeElement clazz, List<VariableElement> fields)
...
+                        fieldAlreadyPresentInSuperClassFormat,
+                        simpleName(InternalId.class),
+                        clazz.getQualifiedName(),
+                        superClazz.getQualifiedName()
+                ));
+            }
+        }
+    }
+
+    private Class<? extends Annotation>[] incompatibleSchemaClassAnnotations(Class<? extends Annotation>... compatibleAnnotations) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java:1207]: We can, but then it will be necessary to do `.toArray(Class[]::new)` everywhere, this is a private method, I suggest leaving it that way, wdyt?

File: modules/configuration-annotation-processor/src/integrationTest/java/org/apache/ignite/internal/configuration/processor/ItConfigurationProcessorTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration-annotation-processor/src/integrationTest/java/org/apache/ignite/internal/configuration/processor/ItConfigurationProcessorTest.java:439]: no, I don't insist

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java:1207]: This method is only called as a parameter of `checkIncompatibleClassAnnotations`, so I think it can be called from there instead. But yes, you will then have to call `toArray` anyway, so my original comment can be considered obsolete

File: modules/configuration-annotation-processor/src/integrationTest/java/org/apache/ignite/internal/configuration/processor/ItConfigurationProcessorTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/integrationTest/java/org/apache/ignite/internal/configuration/processor/ItConfigurationProcessorTest.java:439]: Then I would like to leave it as is.

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java:1207]: Ok

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java [hunk 16]
```diff
@@ -18,24 +18,26 @@
...
-    /** Private constructor. */
-    private Utils() {
-    }
-
+class ConfigurationProcessorUtils {
     /**
      * Get {@link ClassName} for configuration class' public interface.
      *
      * @param schemaClassName Configuration schema ClassName.
-     * @return Configuration's public interface ClassName.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:38]: They duplicate information from the description, so I think they can be removed.
Would you like me to return them?

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java [hunk 17]
```diff
@@ -18,24 +18,26 @@
...
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.element.VariableElement;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Annotation processing utilities.
  */
-public class Utils {
-    /** Private constructor. */
-    private Utils() {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:31]: It is not needed, it is not a public class for users.
Would you like me to return them?

File: modules/configuration-annotation-processor/src/main/resources/META-INF/services/javax.annotation.processing.Processor [hunk 18]
```diff
@@ -1 +1 @@
-org.apache.ignite.internal.configuration.processor.Processor
\ No newline at end of file
+org.apache.ignite.internal.configuration.processor.ConfigurationProcessor
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/resources/META-INF/services/javax.annotation.processing.Processor:1]: Fix it

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+/**
+ * This annotation marks the class as an abstract configuration schema. Has basically the same properties as a {@link PolymorphicConfig},
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java:28]: Fix it

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+ * but its type cannot be changed and its inheritors must contain either {@link Config} or {@link ConfigurationRoot} and also cannot be used
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java:29]: Fix it

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+ * as a nested configuration only its inheritors.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java:30]: That a schema with this annotation cannot be used as a nested configuration, only its descendants can be used.

For example:
`@AbstractConfiguration
public static class AbstractConfigurationSchema {
    @Value(hasDefault = true)
    public int intVal = 100500;
}

@Config
public static class ConfigFromAbstractConfigurationSchema extends AbstractConfigurationSchema {
    @Value(hasDefault = true)
    public boolean booleanVal = true;
}

///

@Config
public static class SuperConfigurationSchema {
	// Error.
    @ConfigValue
    public boolean AbstractConfigurationSchema err;
	
	// Ok.
    @ConfigValue
    public boolean ConfigFromAbstractConfigurationSchema good;
}`

File: modules/configuration/README.md [hunk 22]
```diff
@@ -76,7 +106,9 @@ public static class FirstPolymorphicInstanceConfigurationSchema extends Polymorp
 * `@Config` is similar to the `@ConfigurationRoot` but represents an inner configuration node;
 * `@PolymorphicConfig` is similar to the `@Config` and an abstract class in java, i.e. it cannot be instantiated, but it can be subclassed;
 * `@PolymorphicConfigInstance` marks an inheritor of a polymorphic configuration. This annotation has a single property called `value` - 
-   a unique identifier among the inheritors of one polymorphic configuration, used to define the type (schema) of the polymorphic configuration we are dealing with now.
+   a unique identifier among the inheritors of one polymorphic configuration, used to define the type (schema) of the polymorphic configuration we are dealing with now;
+* `@AbstractConfiguration` is similar to the `@PolymorphicConfig` but its type cannot be changed and its inheritors must contain
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration/README.md:110]: Fix it

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java (hunk 1 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java:3430]: Renamed to `extendsAbstractConfiguration`

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java (hunk 2 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java:3444]: Fix it in IGNITE-17167

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 4 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:1158]: Yes, it makes it shorter, if we remove it, then similar methods should also be removed, I suggest leaving it as it is wdyt?

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 5 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:633]: They duplicate information from the description.
Do you want me to return them?

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java (hunk 6 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:215]: You're right, I'll rename them.

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java (hunk 7 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:224]: Then more memory will be used.

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java (hunk 8 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:248]: Then more memory will be used.

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java (hunk 9 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:258]: Then more memory will be used.

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java (hunk 10 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:257]: Less memory will be used.

File: modules/core/src/test/java/org/apache/ignite/internal/util/CollectionUtilsTest.java (hunk 11 above)
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/util/CollectionUtilsTest.java:278]: Fix it.

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java [hunk 23]
```diff
@@ -71,24 +71,49 @@ public static ClassName getChangeName(ClassName schemaClassName) {
...
+    }
+
+    /**
+     * Returns the first annotation found for the class.
+     *
+     * @param clazz Class type.
+     * @param annotationClasses Annotation classes that will be searched for the class.
+     */
+    @SafeVarargs
+    public static @Nullable Annotation findFirst(
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:100]: Yes, your suggestions sound reasonable.

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java]: What do you think if you say so and replace it with:
`or {@link ConfigurationRoot}. Configuration schemas with this annotation cannot be used as a nested (sub)configuration.`

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:31]: yes, I don't understand why it was removed

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:38]: If they are already written, then I don't see any point of removing them

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java]: yes, that's perfect

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 4 above)
Comment (sashapolo) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:1158]: I would suggest removing this particular one, we can remove others during a separate refactoring

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 5 above)
Comment (sashapolo) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:633]: yes, let's keep them

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:31]: And why is he needed?

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java (hunk 7 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java]: seriously? I think that we agreed that this is not a very good argument

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java [hunk 24]
```diff
@@ -849,12 +848,12 @@ private void checkIncompatibleClassAnnotations(
         assert clazz.getAnnotation(clazzAnnotation) != null : clazz.getQualifiedName();
         assert !nullOrEmpty(incompatibleAnnotations);
 
-        Annotation incompatible = findFirst(clazz, incompatibleAnnotations);
+        Optional<? extends Annotation> incompatible = findFirstPresentAnnotation(clazz, incompatibleAnnotations);
 
-        if (incompatible != null) {
+        if (incompatible.isPresent()) {
```
Comment (sashapolo) [inline modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java:853]: I recommend using `orElseThrow`

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java [hunk 25]
```diff
@@ -1018,7 +1017,7 @@ private void validateInjectedNameFields(TypeElement clazz, List<VariableElement>
             ));
         }
 
-        if (findFirst(clazz, Config.class, PolymorphicConfig.class, AbstractConfiguration.class) == null) {
+        if (findFirstPresentAnnotation(clazz, Config.class, PolymorphicConfig.class, AbstractConfiguration.class).isEmpty()) {
```
Comment (sashapolo) [inline modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java:1020]: I recommend using `orElseThrow`

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java (hunk 7 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java]: What is the best option you suggested?

File: modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-api/src/main/java/org/apache/ignite/configuration/annotation/AbstractConfiguration.java]: Fix it

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:38]: I guess for private API they can be removed if you really want to, I just don't understand why did you have to do it in this PR. Looks like extra unnecessary work

File: modules/configuration/README.md [hunk 26]
```diff
@@ -76,7 +106,9 @@ public static class FirstPolymorphicInstanceConfigurationSchema extends Polymorp
...
+  either `@Config` or `@ConfigurationRoot` and also cannot be used as a nested configuration only its inheritors;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration/README.md:111]: Fix it

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:31]: it prevents you from accidentally instantiating this class and is considered a good practice. Anyway, this code was present before and is not related to your PR. I don't see any positive improvements regarding this change

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:38]: Yes, I want to delete it, it was a joy for me, we leave it as it is.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 5 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:633]: Yes, I want to delete.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 4 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:1158]: It is better to deal with all in one ticket so that there is one style if needed.
https://issues.apache.org/jira/browse/IGNITE-17166

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:31]: It saves you just 3 lines of code and it makes sense

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 4 above)
Comment (sashapolo) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:1158]: but this is a new method that you introduce in this PR, what's the point of leaving it if you are going to remove it later?

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 5 above)
Comment (sashapolo) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:633]: ok =)

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java (hunk 24 above)
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java:853]: It uses `isPresent()` so `orElseThrow` won't work

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessorUtils.java:31]: Discussed in person, decided to remove it.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 4 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:1158]: Fix it.

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java (hunk 24 above)
Comment (sashapolo) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/ConfigurationProcessor.java:853]: you are correct

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java (hunk 7 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java]: Fix it

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java (hunk 9 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java]: Fix it

**Review by sashapolo — APPROVED** [review]
